File: packages/expo-av/android/src/main/java/expo/modules/av/player/datasource/CustomHeadersOkHttpDataSourceFactory.java
package expo.modules.av.player.datasource;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.exoplayer2.ext.okhttp.OkHttpDataSource;
import com.google.android.exoplayer2.upstream.DataSource;
import com.google.android.exoplayer2.upstream.HttpDataSource;
import com.google.android.exoplayer2.upstream.TransferListener;

import java.util.Map;

import okhttp3.CacheControl;
import okhttp3.Call;

// Mainly a copy of com.google.android.exoplayer2.ext.okhttp.OkHttpDataSourceFactory,
// because it's declared as final :(
public class CustomHeadersOkHttpDataSourceFactory extends HttpDataSource.BaseFactory {
  @NonNull
  private final Call.Factory mCallFactory;
  @Nullable
  private final String mUserAgent;
  @Nullable
  private final CacheControl mCacheControl;

  public CustomHeadersOkHttpDataSourceFactory(@NonNull Call.Factory callFactory, @Nullable String userAgent, @Nullable Map<String, Object> requestHeaders) {
    super();
    mCallFactory = callFactory;
    mUserAgent = userAgent;
    mCacheControl = null;
    updateRequestProperties(getDefaultRequestProperties(), requestHeaders);
  }

  protected void updateRequestProperties(HttpDataSource.RequestProperties requestProperties, @Nullable Map<String, Object> requestHeaders) {
    if (requestHeaders != null) {
      for (Map.Entry<String, Object> headerEntry : requestHeaders.entrySet()) {
        if (headerEntry.getValue() instanceof String) {
          requestProperties.set(headerEntry.getKey(), (String) headerEntry.getValue());
        }
      }
    }
  }

  protected OkHttpDataSource createDataSourceInternal(HttpDataSource.RequestProperties defaultRequestProperties) {
    return new OkHttpDataSource(mCallFactory, mUserAgent, null, mCacheControl, defaultRequestProperties);
  }
}

File: packages/expo-av/android/src/main/java/expo/modules/av/player/datasource/CustomHeadersOkHttpDataSourceFactory.kt
package expo.modules.av.player.datasource

import androidx.collection.ArrayMap
import com.google.android.exoplayer2.upstream.HttpDataSource.BaseFactory
import com.google.android.exoplayer2.upstream.HttpDataSource.RequestProperties
import com.google.android.exoplayer2.ext.okhttp.OkHttpDataSource
import okhttp3.CacheControl
import okhttp3.Call

// Mainly a copy of com.google.android.exoplayer2.ext.okhttp.OkHttpDataSourceFactory,
// because it's declared as final :(
class CustomHeadersOkHttpDataSourceFactory(
  private val callFactory: Call.Factory,
  private val userAgent: String?,
  requestHeaders: Map<String, Any>?
) : BaseFactory() {
  private val cacheControl: CacheControl? = null

  private fun updateRequestProperties(requestHeaders: Map<String, Any>?) {
    if (requestHeaders != null) {
      val requestProperties = ArrayMap<String, String>()
      for ((key, value) in requestHeaders) {
        if (value is String) {
          requestProperties[key] = value
        }
      }
      setDefaultRequestProperties(requestProperties)
    }
  }

  override fun createDataSourceInternal(defaultRequestProperties: RequestProperties): OkHttpDataSource {
    return OkHttpDataSource(callFactory, userAgent, cacheControl, defaultRequestProperties)
  }

  init {
    updateRequestProperties(requestHeaders)
  }
}
